Improve readability of decbinhex demo

insomniacUNDERSCORElemon opened the pull request:
* improve icon

Outline, looks better at low resolution

* Finish node type changes

Nodes properly stored in class instead of string for better readability, code condensing

insomniacUNDERSCORElemon committed:
improve icon (#2)

* improve icon

Outline, looks better at low resolution

* Finish node type changes

Nodes properly stored in class instead of string for better readability, code condensing

panno8M commented:
From the viewpoint of readability, it is not a good idea to reduce the number of lines without thought.
This is because source code sometimes functions as a table.
For example, writing a variable definition on a single line is clearly excessive. Code that uses semicolons should not be written too often.

insomniacUNDERSCORElemon committed:
remove 2 of the line breaks from convert.nim

insomniacUNDERSCORElemon commented:
> From the viewpoint of readability, it is not a good idea to reduce the number of lines without thought. This is because source code sometimes functions as a table. For example, writing a variable definition on a single line is clearly excessive. Code that uses semicolons should not be written too often.

I did put thought into context and control flow of the lines. For instance actions that have a similar purpose (resetting values, or an error message and returning, setting window bounds, setting up similar elements). In some cases it makes it more obvious that adjacent lines are doing the same thing, or it creates less scope confusion.

The lines with something like `= (true, true)` I would've preferred if I could've just set them to 1 value, but I just opted to use a different style for differing values to make it more obvious.

panno8M commented on decbinhex4/addons/main/src/code/convert.nim:
```diff
@@ -2,28 +2,24 @@ import math, strutils
-  var result_decimal: int = 0
+proc convert_from_hexadecimal*(input_hex:string):string =
+  var result_binary:string = "reset binary result"; var result_decimal:int = 0
+  var lengthvar_hex:int = input_hex.len-1
```
Let me suggest the proposal for variable definitions using this section.
I think this coding style will also keep the semantic cohesion:

```nim
var
  result_binary: string = "reset binary result"
  result_decimal: int = 0
var
  lengthvar_hex: int = input_hex.len-1
```

insomniacUNDERSCORElemon replied:
The combined line is doing the same thing, putting further emphasis on the line that isn't combined.

However, I believe it may be clearer if I don't assign values:

`var result_decimal: int; var result_hex:string`

Those values should never be visible anyways, and they don't need a value at that time either. EDIT: did that

panno8M replied:
If you really do not want multiple lines, try the following:

```nim
var (result_binary, result_decimal): (string, int) = ("reset binary result", 0)
var lengthvar_hex: int = input_hex.len-1
```

insomniacUNDERSCORElemon replied:
I just changed the line that used the same types to 

`  var result_binary, result_hex:string`.

I already changed the other 2 lines in that file to only initialize.

panno8M commented:
It's hard to understand if you don't use git regularly, but there is a problem that it's hard to track changes if you put similar information on a single line.
An extreme example would be as follows:

```diff
- var a: string = "abc"; var b: string = "def"; var c: string = "Ghi"
+ var a: string = "abc"; var b: string = "def"; var c: string = "ghi"
```

```diff
  var
    a: string = "abc"
    b: string = "def" 
-   c: string = "Ghi" 
+   c: string = "ghi"
```

panno8M commented:
Indeed, the disadvantage of difficult to understand scopes is a potential drawback of python-style languages.
I don't know what editor you are using, but most editors have an extension that colors spaces according to indentation level.
I use vscode and am a regular user of indent-rainbow.

insomniacUNDERSCORElemon committed:
change line in convert.nim

assigning value here is not needed

insomniacUNDERSCORElemon committed:
change other similar lines too, convert.nim

insomniacUNDERSCORElemon committed:
fix convert.nim

I am tired

panno8M commented on decbinhex4/addons/main/src/code/convert.nim:
```diff
@@ -3,22 +3,22 @@ import std/times #benchmark only
 # the asterisk is so it can be called from another file when imported
 proc convert_from_decimal*(input_decimal: string):string =
-  var result_binary: string = "reset binary result"; var result_hex: string = "reset hex result"
+  var result_binary; var result_hex:string
```
This change causes compile error. Because the type cannot be deduced.

panno8M commented on decbinhex4/addons/main/src/code/main_class.nim:
```diff
@@ -1,89 +1,60 @@
-  in_node:string = "grid/inbox"
-  out_node:string = "grid/a/outlabel"
+import strutils, convert #convert.nim
+import gdext, gdextgen/classes / [gdSceneTree, gdInput, gdLineEdit, gdLabel, gdBaseButton], gdextgen/classes / [gdWindow, gdTime]
```
Suggest either:
```nim
import gdext, gdextgen/classes / [gdSceneTree, gdInput, gdLineEdit, gdLabel, gdBaseButton, gdWindow, gdTime]
```

```nim
import gdext
import gdextgen/classes / [gdSceneTree, gdInput, gdLineEdit, gdLabel, gdBaseButton, gdWindow, gdTime]
```

insomniacUNDERSCORElemon replied:
First line of imports is pure nim, second is godot with the ordering denoting importance. `[gdWindow, gdTime]` is not strictly required for the functionality of the program. EDIT: and I just removed gdTime because I didn't use that

insomniacUNDERSCORElemon committed:
(string, string) convert.nim

insomniacUNDERSCORElemon committed:
undo last commit, convert.nim

insomniacUNDERSCORElemon committed:
remove gdTime import main_class.nim

Unneeded, needed functionality was instead imported from the with pure nim file

insomniacUNDERSCORElemon committed:
Update main_class.nim

insomniacUNDERSCORElemon committed:
var result_binary, result_hex:string convert.nim

panno8M commented on decbinhex4/addons/main/src/code/convert.nim:
```diff
@@ -2,28 +2,24 @@ import math, strutils
+  if input_decimal.len < 20: result_binary = strip((input_decimal.parseInt).toBin(63), trailing = false, chars = {'0'})
   else: result_binary = "Output binary too large!"
   result_hex = strip(input_decimal.parseInt.toHex, trailing = false, chars = {'0'})
   return("Binary is: " & result_binary & "\nHexadecimal is: " & result_hex)
```
Is there a reason why you use var at the head of the definition in situations where let can be used?

```nim
proc convert_from_decimal*(input_decimal: string):string =
  let result_binary: string =
    if input_decimal.len < 20:
      strip((input_decimal.parseInt).toBin(63), trailing = false, chars = {'0'})
    else:
      "Output binary too large!"
  let result_hex: string = strip(input_decimal.parseInt.toHex, trailing = false, chars = {'0'})
  return("Binary is: " & result_binary & "\nHexadecimal is: " & result_hex)
```

insomniacUNDERSCORElemon replied:
You're right some of those weren't needed and I changed them now. Didn't think about it, I guess.

Others are either accessed multiple times and/or directly used as storage, like result_decimal.

So now those instances of doubled lines are no longer an issue.

I can't seem to change the first result_binary to let, I assume it's not allowed to initialize variables when branching?

insomniacUNDERSCORElemon committed:
change some var to let, convert.nim

panno8M commented:
I am sorry. I still don't understand the point of going to the trouble of cutting line breaks to make a line longer than 80 characters.
I'm not saying it's hard to read now, but isn't it hard to work with? Doesn't it break lines or overflow on its own?

insomniacUNDERSCORElemon commented:
> I am sorry. I still don't understand the point of going to the trouble of cutting line breaks to make a line longer than 80 characters.
> I'm not saying it's hard to read now, but isn't it hard to work with? Doesn't it break lines or overflow on its own?

My editor is 171 characters wide. It's maximized at 1080p (and a small-ish monitor), default zoom level, 12pt font. Even if I open the sidepanel, it's 150 characters wide. Is that abnormal?

EDIT: Though I don't use a tiling WM (and don't really care to have a window on each half of the screen), so I could see that being a reason for the workflow difference

panno8M commented:
Frankly, I'm still not convinced in the discussion of lines, but this is your production. A big thank you  for your contribution.

insomniacUNDERSCORElemon commented:
> Frankly, I'm still not convinced in the discussion of lines, but this is your production. A big thank you for your contribution.

I have reworked it a bit (convert.nim particularly can now fix almost entirely in a half-editor, max with is 82 characters) but as I was preparing to make a PR I noticed that some things are broken.

It seems to be a larger issue, I just reported it in the gdextshell repo. https://github.com/godot-nim/gdextshell/issues/25
